lib/bloom/sbbf.ts

```diff
@@ -0,0 +1,396 @@
+
+        return Math.ceil(m / this.BITS_PER_BLOCK)
+    }
+
+    /**
+     * @function mask
+     * @description generate a mask block for a bloom filter block
+     * @param hashValue: the hash value to generate the mask from
+     * @private
+     *
+     * @return mask Block
+     */
+    static mask(hashValue: Long): Block {
+        let result: Block = SplitBlockBloomFilter.initBlock()
+        for (let i = 0; i < result.length; i++) {
+            const y = hashValue.getLowBitsUnsigned() * SplitBlockBloomFilter.salt[i]
+            result[i] = result[i] | (1 << (y >>> 27))
+        }
+        return result
+    }
+
+    /**
+     * @function blockInsert
+     * @description insert a hash into a Bloom filter Block
+     * @param b: the block to flip a bit for: is changed
+     * @param hashValue: the hash value to insert into b
+     * @private
+     *
+     * @return void
+     */
+    static blockInsert(b: Block, hashValue: Long): void {
+        const masked: Block = this.mask(hashValue)
+        for (let i = 0; i < masked.length; i++) {
+            for (let j = 0; j < this.WORD_SIZE; j++) {
+                const isSet = masked[i] & (2 ** j)
+                if (isSet) {
+                    b[i] = b[i] | (2 ** j)
+                }
+            }
+        }
+    }
+
+    /**
+     * @function blockCheck
+     * @description check if a hashValue exists for this filter
+     * @param b: the block to check for inclusion
+     * @param hashValue: the hash to check for  should be long
+     * @private
+     *
+     * @return true if hashed item is __probably__ in the data set represented by this filter
+     * @return false if it is __definitely not__ in the data set.
+     */
+    static blockCheck(b: Block, hashValue: Long): boolean {
+        const masked: Block = this.mask(hashValue)
+        for (let i = 0; i < masked.length; i++) {
+            for (let j = 0; j < this.WORD_SIZE; j++) {
+                const isSet = masked[i] & (2 ** j)
+                if (isSet) {
+                    const match = b[i] & (2 ** j)
+                    if (!match) {
+                        return false
+                    }
+                }
+            }
+        }
+        return true
+    }
+
+    /**
+     * Instance
+     */
+
+    private splitBlockFilter: Array<Block> = []
+    private desiredFalsePositiveRate: number = SplitBlockBloomFilter.DEFAULT_FALSE_POSITIVE_RATE
+    private numBlocks: number = 0
+    private numDistinctValues: number = SplitBlockBloomFilter.DEFAULT_DISTINCT_VALUES
+    private hashStrategy = new parquet_thrift.BloomFilterHash(parquet_thrift.XxHash)
+
+    private isInitialized(): boolean { return this.splitBlockFilter.length > 0 }
+
+    getFalsePositiveRate(): number { return this.desiredFalsePositiveRate }
+    getNumDistinct(): number { return this.numDistinctValues }
+    getNumFilterBlocks(): number { return this.splitBlockFilter.length }
+    getFilter(): Array<Block> { return this.splitBlockFilter }
```

Comment:
This feels like it might be a use case for getters and setters, but likely not worth it:

https://www.typescriptlang.org/docs/handbook/2/classes.html#getters--setters

Reply:
eh, if I'm going to make changes anyway, may as well go big TS or go home

Reply:
Maybe another time